Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs

```diff
@@ -142,37 +168,89 @@ public override void OnActionReceived(float[] vectorAction)
         bpDict[forearmR].SetJointStrength(vectorAction[++i]);
     }
 
+    void UpdateOrientationCube()
+    {
+        //FACING DIR
+        m_WalkDir = target.position - m_OrientationCube.transform.position;
+        m_WalkDir.y = 0; //flatten dir on the y
+        m_WalkDirLookRot = Quaternion.LookRotation(m_WalkDir); //get our look rot to the target
+        
+        //UPDATE ORIENTATION CUBE POS & ROT
+        m_OrientationCube.transform.position = hips.position;
+        m_OrientationCube.transform.rotation = m_WalkDirLookRot;
+    }
+    
     void FixedUpdate()
     {
+        if (detectTargets)
+        {
+            foreach (var bodyPart in m_JdController.bodyPartsDict.Values)
+            {
+                if (bodyPart.targetContact && bodyPart.targetContact.touchingTarget)
+                {
+                    TouchedTarget();
+                }
+            }
+        }
+        
+        UpdateOrientationCube();
+        
         // Set reward for this step according to mixture of the following elements.
         // a. Velocity alignment with goal direction.
         // b. Rotation alignment with goal direction.
         // c. Encourage head height.
-        // d. Discourage head movement.
-        m_DirToTarget = target.position - m_JdController.bodyPartsDict[hips].rb.position;
         AddReward(
-            +0.03f * Vector3.Dot(m_DirToTarget.normalized, m_JdController.bodyPartsDict[hips].rb.velocity)
-            + 0.01f * Vector3.Dot(m_DirToTarget.normalized, hips.forward)
-            + 0.02f * (head.position.y - hips.position.y)
-            - 0.01f * Vector3.Distance(m_JdController.bodyPartsDict[head].rb.velocity,
-                m_JdController.bodyPartsDict[hips].rb.velocity)
+            +0.02f * Vector3.Dot(m_OrientationCube.transform.forward,
```

Comment:
To clarify this, I suggest naming each component of this reward with a variable.

```csharp
var hipVelocityReward = Vector3.Dot(m_OrientationCube.transform.forward, Vector3.ClampMagnitude(m_JdController.bodyPartsDict[hips].rb.velocity, maximumWalkingSpeed));
var headOrientationReward = Vector3.Dot(m_OrientationCube.transform.forward, head.forward);
var headHeightOverFeetReward = (head.position.y - footL.position.y) + (head.position.y - footR.position.y);

AddReward(
   0.02f * hipVelocityReward + 
   0.01f * headOrientationReward + 
   0.005f * headHeightOverFeetReward
   );
```

Reply:
agreed that looks cleaner. refactored based on your code.